### PR TITLE
lib: Implement a file selecting widget

### PR DESCRIPTION
**Pull request by petervo**
Give users a little autocomplete help when they need to
select a file from the server. We need this for upcoming
ssh keys and VM dialogs.

### CODE REVIEW

**Comment by andreasn** [comment]
This is awesome!
Works really great in general.

If I select something in the dropdown menu with the mouse, I'll end up with double slashes at the end.
screenshot from 2017-07-04 11-46-13

The height of the dropdown list can get so long it scrolls outside the browser height.
Not sure what's the best height here. Maybe around 15 rows?
screenshot from 2017-07-04 11-45-35

**Comment by petervo** [comment]
@andreasn fixed those two issues.

File: pkg/lib/cockpit-components-file-autocomplete.jsx [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+            listItems = React.Children.map(this.state.displayFiles, function(file) {
+                return <li className={file.type}><a data-type={file.type}>{file.path}</a></li>;
+            });
+        }
+
+        return (
+            <div className="combobox-container" id={this.props.id}>
+                <div className={classes}>
+                    <input autocomplete="false" placeholder={placeholder} className="combobox form-control" type="text" onChange={this.delayedOnChange} value={this.state.value} />
+                    <span title="lala" onClick={this.showAllOptions} className={controlClasses}></span>
```
Comment (andreasn) [inline pkg/lib/cockpit-components-file-autocomplete.jsx]: title="lala" looks odd. Leftover?

**Comment by andreasn** [comment]
Showed it briefly to @fabianapsimoes and her feedback was that it worked great, but that the error when you mistype a path was a bit too intense, but we can solve that as a followup issue.

File: pkg/lib/cockpit-components-file-autocomplete.jsx (hunk 1 above)
Comment (petervo) [inline pkg/lib/cockpit-components-file-autocomplete.jsx]: good catch, fixed.

**Comment by andreasn** [comment]
> Showed it briefly to @fabianapsimoes and her feedback was that it worked great, but that the error when you mistype a path was a bit too intense, but we can solve that as a followup issue.

Fixed that here: https://github.com/andreasn/cockpit/tree/file-widget-error

**Comment by petervo** [comment]
Updated the error state display

File: pkg/lib/cockpit-components-file-autocomplete.jsx [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        if (this.state.error)
+            error = cockpit.format(cockpit.message(this.state.error));
+        else if (this.state.directoryFiles && this.state.displayFiles.length < 1)
+            error = _("No matching files found");
+
+        if (error) {
+            listItems = [this.renderError(error)];
+            classes += " error"
+        } else {
+            listItems = React.Children.map(this.state.displayFiles, function(file) {
```
Comment (dperpeet) [inline pkg/lib/cockpit-components-file-autocomplete.jsx]: I think there should be some limitation in place regarding the number of entries, so the ui doesn't freeze. It doesn't have to include pagination. In a first step I would set a hard limit, e.g. 500.

Comment (petervo) [inline pkg/lib/cockpit-components-file-autocomplete.jsx]: 500 is actually pretty low it deals with that pretty well. I set it at 5000.

File: pkg/lib/cockpit-components-file-autocomplete.jsx [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+                open: false,
+            });
+            this.updateFiles(directory);
+        }
+        return changed;
+    },
+
+    finishUpdate: function(result, error) {
+        result = result.sort(function(a, b) {
+            return a.path.localeCompare(b.path,
```
Comment (dperpeet) [inline pkg/lib/cockpit-components-file-autocomplete.jsx]: technically we should set the locale here, but I don't think it matters really

File: pkg/lib/cockpit-components-file-autocomplete.jsx (hunk 2 above)
Comment (dperpeet) [inline pkg/lib/cockpit-components-file-autocomplete.jsx]: We could have shown those 5000 and elided the rest, but that could also be confusing... this is good.

**Comment by dperpeet** [comment]
There are some possible usability improvements, such as not closing the dropdown if a path is selected - it could show the next level or even tab completion. But all that is good for a follow-up if necessary. Yay for having this!

**Pull request merged**